File: frontend/js/views/hsa_triple_tax.js
// HSA triple-tax advantage — HSA vs a taxable account over a horizon, the
// dollar value of deductible-in / tax-free-growth / tax-free-out, via
// /calc/hsa-triple-tax. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const FIELDS = [
    ['annual_contribution_usd', 'Annual contribution ($)', 4000],
    ['years', 'Years', 20],
    ['annual_growth_pct', 'Annual growth (%)', 7],
    ['marginal_tax_rate_pct', 'Marginal tax rate (%)', 24],
    ['ltcg_rate_pct', 'Long-term cap-gains rate (%)', 15],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const VIEW = 'hsa-triple-tax';
let lastReport = null;
let lastBody = null;

export async function renderHsaTripleTax(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.hsatt.h1.title">// HSA TRIPLE-TAX ADVANTAGE</span></h1>
        <p class="muted small" data-i18n="view.hsatt.hint.intro">
            The HSA is the only account taxed favorably three ways: contributions are
            deductible, growth is tax-free, and withdrawals for medical expenses are tax-free.
            This projects an HSA against a taxable account funded with the same gross income —
            the taxable side invests only the after-tax amount and pays cap-gains on its growth.
            The gap is the dollar value of the triple-tax treatment. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.hsatt.h2.inputs">The plan</h2>
            <form id="hsa-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.hsatt.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="hsa-tools" class="ce-toolbar"></div>
            <button type="button" id="hsa-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="hsa-sens" class="ce-sens"></div>
        </div>
        <div id="hsa-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#hsa-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        body.years = Math.max(0, Math.round(body.years));
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcHsaTripleTax(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.hsatt.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#hsa-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'hsa-triple-tax.csv' });
    mount.querySelector('#hsa-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['hsa_advantage_usd', r.hsa_advantage_usd],
        ['hsa_ending_usd', r.hsa_ending_usd],
        ['taxable_ending_usd', r.taxable_ending_usd],
        ['total_contributions_usd', r.total_contributions_usd],
        ['upfront_tax_savings_usd', r.upfront_tax_savings_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#hsa-result');
    // Line chart: HSA advantage as the horizon sweeps 0 -> 40 years (compounding widens the gap).
    const xs = enh.linspace(0, 40, 13);
    const pts = await Promise.all(xs.map(async (yr) => {
        const rr = await api.calcHsaTripleTax({ ...body, years: Math.round(yr) });
        return { x: yr, y: rr ? rr.hsa_advantage_usd / 1000 : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'years', ylabel: 'advantage $k' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.hsatt.h2.result">The advantage</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.hsatt.card.advantage">HSA advantage</div>
                    <div class="value pos">${money(r.hsa_advantage_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.hsatt.card.hsa">HSA ending value</div>
                    <div class="value">${money(r.hsa_ending_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.hsatt.card.taxable">Taxable ending value</div>
                    <div class="value">${money(r.taxable_ending_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <thead><tr><th data-i18n="view.hsatt.col.line">Line</th><th data-i18n="view.hsatt.col.amount">Amount</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.hsatt.row.contributions">Total contributions</td><td>${money(r.total_contributions_usd)}</td></tr>
                    <tr><td data-i18n="view.hsatt.row.upfront">Upfront tax savings (deduction)</td><td>${money(r.upfront_tax_savings_usd)}</td></tr>
                    <tr><td data-i18n="view.hsatt.row.hsa">HSA ending (tax-free)</td><td>${money(r.hsa_ending_usd)}</td></tr>
                    <tr><td data-i18n="view.hsatt.row.taxable">Taxable ending (after cap-gains)</td><td>${money(r.taxable_ending_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.hsatt.row.advantage">Triple-tax advantage</td><td class="pos">${money(r.hsa_advantage_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#hsa-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: years 0 -> 40; y: annual growth 2% -> 12%. Output: HSA advantage.
    const xVals = enh.linspace(0, 40, 5);
    const yVals = enh.linspace(2, 12, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'years', yKey: 'annual_growth_pct', xVals: xVals.map(Math.round), yVals, compute: (b) => api.calcHsaTripleTax(b), pick: (r) => (r ? r.hsa_advantage_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals: xVals.map(Math.round), yVals, cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v / 1000) + 'k'), xfmt: (v) => v.toFixed(0) + 'y', yfmt: (v) => v.toFixed(0) + '%', xName: t('view.hsatt.label.years') || 'Years', yName: t('view.hsatt.label.annual_growth_pct') || 'Growth' });
}

File: crates/traderview-core/src/hsa_triple_tax.rs
//! HSA triple-tax advantage — HSA vs a taxable account over time.
//!
//! The HSA is the only account taxed favorably three ways: contributions are
//! deductible (pre-tax), growth is tax-free, and withdrawals for qualified
//! medical expenses are tax-free. This quantifies that edge by projecting an
//! HSA against a taxable brokerage account funded with the same gross income:
//!
//!   * **HSA**: the full contribution goes in pre-tax, compounds tax-free, and
//!     comes out tax-free → ending value = the annuity future value.
//!   * **Taxable**: the same pre-tax dollars are taxed first (so only
//!     `contribution × (1 − marginal)` is invested), compound, and the gains
//!     are taxed at the long-term rate at the end.
//!
//! The difference is the dollar value of the triple-tax treatment. Pure
//! compute (end-of-year contributions, gains taxed once at the horizon).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct HsaInput {
    pub annual_contribution_usd: f64,
    pub years: u32,
    pub annual_growth_pct: f64,
    /// Marginal income-tax rate — the deduction value and the taxable
    /// account's haircut on contributions.
    pub marginal_tax_rate_pct: f64,
    /// Long-term cap-gains rate applied to the taxable account's gains.
    pub ltcg_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct HsaResult {
    pub total_contributions_usd: f64,
    /// Upfront tax saved by deducting the contributions.
    pub upfront_tax_savings_usd: f64,
    /// HSA ending value (tax-free in and out).
    pub hsa_ending_usd: f64,
    /// Taxable account ending value, net of cap-gains tax on the gains.
    pub taxable_ending_usd: f64,
    /// HSA − taxable: the dollar value of the triple-tax treatment.
    pub hsa_advantage_usd: f64,
}

/// Future value of `n` end-of-year contributions of `c` growing at rate `r`.
fn annuity_fv(c: f64, r: f64, n: u32) -> f64 {
    if r.abs() < 1e-12 {
        c * n as f64
    } else {
        c * (((1.0 + r).powi(n as i32) - 1.0) / r)
    }
}

pub fn analyze(i: &HsaInput) -> HsaResult {
    let c = i.annual_contribution_usd.max(0.0);
    let r = i.annual_growth_pct / 100.0;
    let marginal = i.marginal_tax_rate_pct / 100.0;
    let ltcg = i.ltcg_rate_pct / 100.0;
    let n = i.years;

    let total_contributions = c * n as f64;
    let upfront_tax_savings = total_contributions * marginal;

    // HSA: full pre-tax contribution, tax-free throughout.
    let hsa_ending = annuity_fv(c, r, n);

    // Taxable: the same gross dollars are taxed before investing.
    let taxable_contrib = c * (1.0 - marginal);
    let taxable_fv = annuity_fv(taxable_contrib, r, n);
    let taxable_basis = taxable_contrib * n as f64;
    let taxable_gains = (taxable_fv - taxable_basis).max(0.0);
    let taxable_ending = taxable_fv - taxable_gains * ltcg;

    HsaResult {
        total_contributions_usd: total_contributions,
        upfront_tax_savings_usd: upfront_tax_savings,
        hsa_ending_usd: hsa_ending,
        taxable_ending_usd: taxable_ending,
        hsa_advantage_usd: hsa_ending - taxable_ending,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> HsaInput {
        HsaInput {
            annual_contribution_usd: 4_000.0,
            years: 20,
            annual_growth_pct: 7.0,
            marginal_tax_rate_pct: 24.0,
            ltcg_rate_pct: 15.0,
        }
    }

    #[test]
    fn total_contributions_and_upfront_savings() {
        let r = analyze(&base());
        assert!((r.total_contributions_usd - 80_000.0).abs() < 1e-6); // 4k × 20
        assert!((r.upfront_tax_savings_usd - 19_200.0).abs() < 1e-6); // 80k × 24%
    }

    #[test]
    fn no_growth_hsa_is_just_contributions() {
        let r = analyze(&HsaInput { annual_growth_pct: 0.0, ..base() });
        assert!((r.hsa_ending_usd - 80_000.0).abs() < 1e-6);
        // Taxable: 4k×(1−.24)=3040/yr × 20 = 60,800, no gains taxed.
        assert!((r.taxable_ending_usd - 60_800.0).abs() < 1e-6);
    }

    #[test]
    fn hsa_annuity_fv_with_growth() {
        // 4000 × ((1.07^20 − 1)/0.07).
        let r = analyze(&base());
        let expected = 4_000.0 * ((1.07_f64.powi(20) - 1.0) / 0.07);
        assert!((r.hsa_ending_usd - expected).abs() < 1e-3);
    }

    #[test]
    fn hsa_beats_taxable() {
        let r = analyze(&base());
        assert!(r.hsa_advantage_usd > 0.0);
        assert!(r.hsa_ending_usd > r.taxable_ending_usd);
    }

    #[test]
    fn taxable_gains_are_taxed_at_end() {
        let r = analyze(&base());
        let taxable_contrib = 4_000.0 * 0.76;
        let taxable_fv = taxable_contrib * ((1.07_f64.powi(20) - 1.0) / 0.07);
        let gains = taxable_fv - taxable_contrib * 20.0;
        let expected = taxable_fv - gains * 0.15;
        assert!((r.taxable_ending_usd - expected).abs() < 1e-3);
    }

    #[test]
    fn higher_marginal_rate_widens_advantage() {
        let low = analyze(&HsaInput { marginal_tax_rate_pct: 12.0, ..base() });
        let high = analyze(&HsaInput { marginal_tax_rate_pct: 37.0, ..base() });
        assert!(high.hsa_advantage_usd > low.hsa_advantage_usd);
    }

    #[test]
    fn upfront_savings_scale_with_marginal() {
        let r = analyze(&HsaInput { marginal_tax_rate_pct: 37.0, ..base() });
        assert!((r.upfront_tax_savings_usd - 80_000.0 * 0.37).abs() < 1e-6);
    }

    #[test]
    fn zero_contribution_zero_everything() {
        let r = analyze(&HsaInput { annual_contribution_usd: 0.0, ..base() });
        assert!(r.hsa_ending_usd.abs() < 1e-9);
        assert!(r.taxable_ending_usd.abs() < 1e-9);
        assert!(r.hsa_advantage_usd.abs() < 1e-9);
    }
}
